Signal/src/contact/OWSContactsManager.m

```diff
@@ -240,170 +232,76 @@ - (void)updateSignalAccounts
                     DDLogDebug(@"Ignoring duplicate contact: %@, %@", signalAccount.recipientId, contact.fullName);
                     continue;
                 }
-                signalAccountMap[signalAccount.recipientId] = signalAccount;
                 [signalAccounts addObject:signalAccount];
             }
         }
 
-        NSArray<NSString *> *lastKnownContactRecipientIds = [signalAccountMap allKeys];
-        [TSStorageManager.sharedManager.newDatabaseConnection
-            readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
-                [transaction setObject:lastKnownContactRecipientIds
-                                forKey:kTSStorageManager_lastKnownContactRecipientIds
-                          inCollection:kTSStorageManager_OWSContactsManager];
-            }];
+        // Update cached SignalAccounts on disk
+        [self.dbWriteConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+            NSArray<NSString *> *allKeys = [transaction allKeysInCollection:[SignalAccount collection]];
+            NSMutableSet<NSString *> *orphanedKeys = [NSMutableSet setWithArray:allKeys];
+            
+            for (SignalAccount *signalAccount in signalAccounts) {
+                // TODO only save the ones that changed
+                [orphanedKeys removeObject:signalAccount.uniqueId];
+                [signalAccount saveWithTransaction:transaction];
```

Comment:
I thought you were going to persist the collection as a whole. Why the change of heart?

Reply:
I started that way, but I realized there are still API's to look up the SignalRecipient by ID. If the cache wasn't been loaded when that happens we can produce the correct result by fetching them from disk directly.